SalesforceCommerceCloud/pwa-kit
# SPIKE: Add Multisite support

## Description

GUS ticket: https://gus.lightning.force.com/a07EE000008vKbxYAE

Quip doc https://salesforce.quip.com/jQmBA3F7rjci




## Why
Currently, PWA Kit only supports 1 single site per project. This might be challenging for customers who have many brands and regions around the world because they will have to create a project per site. As a result, they will have a large number of projects and codebases to manage. To tackle that problem, we want to explore the options to allow PWA kits to be able to support multi-site so they can manage all of their sites in a single codebase

There are 4 options available, but in this PR, we focus on **Option 1A** and **Option 2**. 

Option 1a: When a developer is standing up a multi-site project, they can configure a *single* codebase running on a *single* MRT environment/domain to work with multiple sites configured in Commerce Cloud business manager.
For example:
- pwa-kit.mobifystorefront.com/us → RefArch site
- pwa-kit.mobifystorefront.com/global → RefArchGlobal site

Option 2: When a developer is standing up a multi-site project, they can configure a *single* codebase running on *multiple* MRT environments/domains to work with multiple sites configured in Commerce Cloud business manager
For example:
- us.pwa-kit.mobifystorefront.com → RefArch site
- intl.pwa-kit.mobifystorefront.com → RefArchGlobal site

## Assumption
- I would like to assume we will have **the sitealias on the url at all time** regardless of the option because it avoids affecting the flow of any functions that take url as an input. 
For example, we have the `getLocaleUrl` that  finds the locale from the url and replaces it with a new one. If we make siteAlias optional, that function will have to take whether the siteAlias is defined or not into account, and there is not guarantee that the siteAlias can't have the same pattern as locale.  

## Findings
To be able to support multi site for the PWA Kit, we need to provide the [configuration](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/167/files#diff-84f04418e6d59f3b507c533f794b4bae457e63c71f817fa53f46336e3ccede13R2) for it. The configuration needs a `defaultSiteId` as a fallback for option 1A from the home page. The `sites` list will contain the list of site Info for each site includes `id`, `alias`, `hostnames`, `localization` config. 
- The idea is we want to prioritize looking for `siteId` based on `hostnames` first (which indicates it is option 2).  
- If none is found or there are more than two siteID found based on the hostname, this indicates it is option 1a, the siteId will be determined based on the `siteAlias` from the url. (except if it is the home page, default Site Id will be used).
For example:
- When a user first access `pwa-kit.mobifystorefront.com`, since there is no `siteAlias` in the url, the `defaultSiteId` will be used (eg. RefArchGlobal) whose alias is `global`. The rest of the routings should based on that and build routings with the corresponding siteAlias. For example: pwa-kit.mobifystorefront.com/global/womens/dresses
- If they want to switch to another siteId, Assuming there should be a link that directs them to another site. Let's say the link for that button is `pwa-kit.mobifystorefront.com/us`, the `us` in this case is the `siteAlias` for the second siteId.


## Risks
- How do we determine the schema of the configuration file to make sure that it will be stable and not cause any breaking change for future versions?
- Should we enforce the locale configuration stay the same for multi sites? I have the impression it is not, so we need to refactor some of our code to adopt the localization configuration based on sites configuration as well

## Implementation
If we take a closer look
- For Option 2, the siteId depends on the hostname. 
- For Option 1a, it does not depends on the hostname. On the homepage, we need to take in a default siteID value (like what we did for default locale)
Therefore, the solution is to have a utility function `getSiteId` which can return a siteId that first priority on hostname declared in the app config file. If there is no siteId, or more than 1 is found, it will switch to use default value. 
- Aside from siteId, we also need to function to determine the site alias for the url. Just like the `getSiteId`, we need to determine whether it is from `hostname` or from the siteId for option 2 and option 1a respectively. That is where the `useSiteAlias` comes it. It will prioritize finding the alias based on hostname, if it can't find one or the results is more than 1 (which means the multi site is a single domain aka option 1a) , it will use the url as the source to get the alias.

## Changes
- Added A configuration file that contains the sites configuration under a category called "app". This is to indicate this section will be applied for the pwa kit itself. Potentially, we can use this file to add more configuration that is out of the pwa kit like ssr configuration. For example.
```
{
    "ssr": {
        // All the settings for SSR like proxies and node version etc.
    },
    // Anything within app will fall outside of the `pwa-kit-config` specification.
    // This will hold configuration settings that are app specific, and not
    // releated to the sdk. You can think of this as a custom settings section.
    "app": {
        // `sites` will be the configuration setting that will allow use to determine
        // the site id for  a given alias, as well as which local config to use for the SDK
        // as defined by the work in the PR related to adding local support to the SDK.
        "sites": [{
            "id": "RefArch",
            "alias": "us",
            // These `l10n` settings will override those defined at the top level since we
            // can have specific localization support per site.
            "l10n": {
                "defaultLocale": "en-US",
                "supportedLocales": ["en-US"]
            }
        }, {
            "id": "RefArchGlobal",
            "alias": "global",
            "l10n": {
                "defaultLocale": "en-GB",
                "supportedLocales": ["en-GB", "ja-JP", "fr-FR"]
            }
        }]
    }
}

```

## How to Test-Drive This PR
- `npm start` to start the pwa kit
- Access `localhost:300`, navigate around (PLP, PDP).
- Notice that the route should include the siteIDAlias for all routes


## TODO
- A guidance on how to deployment multi site project to Managed Runtime
- A guidance on how the configuration works

## General
